server/base/src/main/java/org/apache/accumulo/server/util/MasterMetadataUtil.java

```diff
@@ -253,4 +253,27 @@ public static void updateTabletDataFile(ServerContext context, KeyExtent extent,
 
     tablet.mutate();
   }
+
+  /**
+   * Last Location update function updates the tablets last location to its current
+   */
+  public static void updateLastLocation(ServerContext context, KeyExtent extent, MetadataTime time,
+      String address, ZooLock zooLock, TServerInstance lastLocation) {
+    
+    TabletMutator tablet = context.getAmple().mutateTablet(extent);
+    // unsure if this time needs to be updated or not. Need input.
+    tablet.putTime(time);
+
+    TServerInstance self = getTServerInstance(address, zooLock);
+    tablet.putLocation(self, LocationType.LAST);
+
+    // remove the old location
+    if (lastLocation != null && !lastLocation.equals(self)) {
+      tablet.deleteLocation(lastLocation, LocationType.LAST);
+    }
+
+    tablet.putZooLock(zooLock);
```

Comment:
General question for now - how expensive is this locking?  Have you run any performance benchmarks?  It may be common to have lots of tablets being (re)assigned on a bulk-ingest, especially if the users provide ingest files that are not optimally split on table boundaries.  (And, is there an unlock needed?, going to look, but thought I'd ask up front)

Reply:
I am not quite sure on how expensive the locking is. I need to look into running more performance tests at it.

Reply:
When you look at performance, you should consider the situation when there is a lot of movement / lots of reassignments occurring. 

And a more general issue -  I think the general question raised by @keith-turner "Did you consider setting the last location when a tablet is loaded on a tserver?" is still open.

Once a tablet is assigned to a tserver, it should stay there until tserver failure or other reassignment, which should naturally update the location on that reassignment.  Is this issue with bulk imports?  If so, it is possible that one file will be assigned to multiple tablets - each serving a portion of the file.

I guess I'm asking for more background on what improvement / issue that this is working to improve so I can put it into context / my mental model.

Reply:
I looked into setting the last location when a tablet is loaded but ran into several errors. Some time has passed since I have looked at it so I will look back into that implementation tomorrow to get more detail on it.

The issue this was trying to solve was lastLocation not being updated after a cluster was balanced and in a steady state. When the cluster was taken down and brought back up, it would have to rebalance which for larger data sets would take a considerable amount of time. This issue was brought up originally in 1.9 with the data set @ivakegg was using. I tested some different implementations for this and the periodic update seemed to work the best.

Reply:
> And a more general issue - I think the general question raised by @keith-turner "Did you consider setting the last location when a tablet is loaded on a tserver?" is still open.

I have a solution that applies the update location on assignment to the tserver that seems to work just as well. It still needs more testing and maybe a few adjustments.